services/st_util/FlashOnStBootloaderCommunicator.cpp

```diff
@@ -0,0 +1,69 @@
+#include "services/st_util/FlashOnStBootloaderCommunicator.hpp"
+#include "infra/util/ByteRange.hpp"
+#include "infra/util/MemoryRange.hpp"
+#include "services/st_util/StBootloaderCommunicator.hpp"
+#include <cstdint>
+
+namespace services
+{
+    FlashOnStBootloaderCommunicatorBase::FlashOnStBootloaderCommunicatorBase(hal::Flash& flash, StBootloaderCommunicator& communicator)
+        : communicator(communicator)
+        , flash(flash)
+        , onWriteMemoryDone([this]()
+              {
+                  WriteBuffer(this->writeTail, this->tailAddress, std::exchange(this->onDone, nullptr));
```

Comment:
if infra::AutoResetFunction<void()> is used for this->onDone, then std::exchange(..., nullptr) is not necessary